File: solutions/2023/11-Cosmic_Expansion/s1.js
const utils = require('../../../templates/javascript/utils.js')
const {submit} = require('../../../aoc-utils/utils.js')
const path = require('path')
const readFileSync = require('fs').readFileSync
const inputPath = path.join(__dirname, 'input_example')
const data = readFileSync(inputPath, 'utf8')

function parseInput(data) {
    return data.split(/\n/)
}
function isRowEmpty(row) {
 return row.split('').every(x=>x==='.')
}
function isColumnEmpty(column) {
    return column.every(x=>x==='.')
}
function expandRow(grid, rowI) {
    grid.splice(rowI, 0, '.'.repeat(grid[0].length))

}
function expandColumn(grid, columnI) {
    for (const rowI in grid) {
        const newRow = grid[rowI].slice(0,columnI) + '.' + grid[rowI].slice(columnI)
        grid[rowI] = newRow
    }
}
// mutates
function expandUniverse(grid) {
    for (let i=0; i<grid.length; i++) {
        if (isRowEmpty(grid[i])) {
            expandRow(grid, i)
            i++
        }
    }

    for (let i=0; i<grid[0].length; i++) {
        const column = []
        for (let y=0; y<grid.length; y++) {
            column.push(grid[y][i])
        }
        if (isColumnEmpty(column)) {
            expandColumn(grid, i)
            i++
        }
    }

    return grid
}
function findGalaxies(grid) {
    const galaxyLocations = []
    for (let y=0; y<grid.length; y++) {
        for (let x=0; x<grid[0].length; x++) {
            if (grid[y][x] === '#') {
                galaxyLocations.push({x,y})
            }
        }
    }
    return galaxyLocations
}
function findShortestPaths(galaxyLocations) {
    const shortestPaths = []
    for (let g1=0; g1<galaxyLocations.length-1; g1++) {
        for (let g2=g1+1; g2<galaxyLocations.length; g2++) {
            shortestPaths.push(Math.abs(galaxyLocations[g2].x-galaxyLocations[g1].x)+Math.abs(galaxyLocations[g2].y-galaxyLocations[g1].y))
        }
    }
    return shortestPaths
}

const grid = parseInput(data)
// for (const row of grid) {
//     console.log(row)
// }
const expanded = expandUniverse(grid)
// for (const row of expanded) {
//     console.log(row)
// }
const galaxyLocations = findGalaxies(expanded)
// console.log(galaxyLocations)
shortestPaths = findShortestPaths(galaxyLocations)
console.log(shortestPaths)
const SOLUTION_1 = shortestPaths.reduce((a,c)=>a+c)

module.exports = {}

// submit(SOLUTION_1)

console.log(`SOLUTION_1: ${SOLUTION_1}`)
// Example solution: 374

File: solutions/2024/15-Warehouse_Woes/s1.py
import os
import sys

current_dir = os.path.dirname(os.path.realpath(__file__))
utils_path = os.path.join(current_dir, "../../../templates/python")
sys.path.append(utils_path)
import utils

# input_file = "input_example"
input_file = "input"

input_path = os.path.join(current_dir, input_file)

with open(input_path, "r", encoding="utf8") as file:
    data = file.read().split("\n\n")
    grid = data[0]
    grid = [
        list(row)
        for row in grid.split("\n")
    ]
    instructions = data[1].replace('\n', '')
    print(instructions)

def solution_1():
    max_x = len(grid[0])
    max_y = len(grid)
    directions = {
        "^": (0, -1),
        "v": (0, 1),
        "<": (-1, 0),
        ">": (1, 0)
    }
    
    def move(xi, yi, dir):
        xn = xi + directions[dir][0]
        yn = yi + directions[dir][1]
        
        if grid[yn][xn] == '.':
            grid[yn][xn] = grid[yi][xi]
            grid[yi][xi] = '.'
            return xn, yn
            
        if grid[yn][xn] == 'O':
            if move(xn, yn, dir) != (xn, yn):
                grid[yn][xn] = grid[yi][xi]
                grid[yi][xi] = '.'
                return xn, yn
            
        return xi, yi
            
    for _x in range(max_x):
        for _y in range(max_y):
            if grid[_y][_x] == "@":
                xr = _x
                yr = _y
                break
            
    for dir in instructions:
        xr, yr = move(xr, yr, dir)
    
    solution = 0
    for _x in range(max_x):
        for _y in range(max_y):
            if grid[_y][_x] == 'O':
                gps_coord = 100 * _y + _x
                solution += gps_coord
    print(solution)
    

def solution_2():
    max_x = len(grid[0])
    max_y = len(grid)
    directions = {
        "^": (0, -1),
        "v": (0, 1),
        "<": (-1, 0),
        ">": (1, 0)
    }
    
    def move(xi, yi, dir):
        xn = xi + directions[dir][0]
        yn = yi + directions[dir][1]
        
        if grid[yn][xn] == '.':
            grid[yn][xn] = grid[yi][xi]
            grid[yi][xi] = '.'
            return xn, yn
            
        if grid[yn][xn] == 'O':
            if move(xn, yn, dir) != (xn, yn):
                grid[yn][xn] = grid[yi][xi]
                grid[yi][xi] = '.'
                return xn, yn
            
        return xi, yi
            
    for _x in range(max_x):
        for _y in range(max_y):
            if grid[_y][_x] == "@":
                xr = _x
                yr = _y
                break
            
    for dir in instructions:
        xr, yr = move(xr, yr, dir)
    
    solution = 0
    for _x in range(max_x):
        for _y in range(max_y):
            if grid[_y][_x] == 'O':
                gps_coord = 100 * _y + _x
                solution += gps_coord
    print(solution)
    

if __name__ == "__main__":
    # solution_1()
    solution_2()